### ISSUE TITLE
You cannot override inputs and outputs from templates if they aren't specified in the target

### ISSUE THREAD

**Issue by mailund** [mailund/gwf#151]
See discussion in issue #150

**Comment by mailund** [comment]
So, we changed the behaviour from earlier, where `inputs` and `outputs` had default options, so people couldn't misspell them and get the default empty lists by accident. The target construction takes all possible keyword options, so this would easily be a source of errors.

Now, you have to provide the options explicitly both in the `Target` constructor

```python
    def __init__(self, name, inputs, outputs, options, working_dir, spec=''):
```

and in the workflow `target` function

```python
    def target(self, name, inputs, outputs, **options):
```

This doesn't work well with overriding them using the left-shift operator, the way we would do with templates.

We do more validation of the options in the `Target` constructor, by the way, since we also check their type there, which will also be a problem if we override them with the left-shift operator.

The only really good place to validate targets is before we analyse the workflow. Until that point, they can be modified. So what about we make a `validate` method for targets and call that somewhere just before we analyse the workflow?

We would have to do more than validation there, though. The way we use `inputs` and `outputs` as properties of the targets, we need to move them from options to the properties as well. We can also do that in the validation method, but that gets a little ugly.

I think we should probably do this for now, to fix the bug @micknudsen has, and then see if we can come up with a nicer solution later. What do you guys think? @dansondergaard ?

**Comment by dansondergaard** [comment]
Seems like a fine temporary solution, let's do it.

I really think we need to rethink the template mechanism at some point. I love that templates are just functions, but the order of assignment/target creation that causes this problem is hell, both from a validation perspective, but also from a code simplicity perspective.

**Comment by mailund** [comment]
Well, you weren't happy about writing a meta-language for templates either ;)

**Comment by dansondergaard** [comment]
That language wasn't for templates, it was for setting options. Totally overkill!

A fix could be to allow something like this:

```
gwf.target('MyTarget', template(foo='bar', zoo='car'), cores=16)
```

Basically, the function has two signatures: the usual one that takes name inputs, outputs, options etc. and this one which takes a name and a "template tuple", along with options that override the template options. This removes the problem since we can now assign everything when the `Target` is initialized. Technically, the target could be immutable :)

**Comment by mailund** [comment]
Immutable targets might be a very good idea for these exact reasons.

I will make an implementation that solves @micknudsen immidiate problems and then let's talk about this after. We would have to change the syntax, which would cause some problems with all workflows currently using this version, though :(

**Comment by dansondergaard** [comment]
Yea, I would love to get it right before 1.0 since we didn't promise any API stability yet. Once we reach 1.0 it becomes a lot more annoying to people if we change this.

**Comment by mailund** [comment]
Yes. We really should find a good solution before 1.0

Incidentally, I just found that we really should copy the options dict -- which I will do in my @micknudsen fix -- because otherwise we *can* modify options after we have validated targets.

This is not something we have ever had to worry about before, but my test cases failed because of it.

**Comment by dansondergaard** [comment]
Maybe something like this:

```python
gwf.target('MyTarget', my_template(arg1=42, arg2='foo'), cores=16)
```

So `target()` must either take inputs and outputs (and some options), or a "template tuple" (and some options). This way, we can initialize the target correctly from the start and we don't have to mutate it, but the user can still override options set by the template (as in the example where `cores=16` overrides the template default).

If we want to be explicit, this could be handled by a `Workflow.from_template` method:

```python
gwf.target('NormalTarget', inputs=['hello']', outputs=['world'], cores=2) << """echo hello world"""
gwf.from_template('MyTarget', my_template(arg1=42, arg2='foo'), cores=16)
```

This would be simple to implement:

```python
def from_template(self, name, template, **kwargs):
    template_options, template_spec = template

    # Maybe raise a nice exceptions if the template didn't declare any inputs/outputs.
    inputs = template_options.pop('inputs')
    outputs = template_options.pop('outputs')

    options = template_options.copy()
    options.update(kwargs)
    
    new_target = Target(
        name, inputs, outputs, options,
        working_dir=self.working_dir,
    )

    self._add_target(new_target)
    return new_target
```

Thoughts?

**Comment by mailund** [comment]
I have given it some thoughts but I'm not sure...

We cannot really make targets immutable since we need to return them from the `gwf.target()` method so they can be given a spec. We don't want to provide the spec as a method option -- the syntax we have now is nicer. We could, of course, chose not to return the target from the `gwf.from_template()` method, but that would make the interface inconsistent.

And, by the way, raising an exception if a template does not provide input and output files is probably not a good idea -- it would be perfectly valid for those to be provided directly to the target and not come from the template.

Would it be possible to handle all this with the `inputs` and `outputs` properties? The `validate()` solution I wrote is ugly because it is not a simple predicate. It does all kinds of stuff that modifies the target as well. We *could* avoid this by having a flag in the properties that tells us if the `inputs` and `outputs` have been set and still have a `validate()` predicate. That way, we can keep the syntax we have now. We would still need to explicitly pop the options for `inputs` and `outputs` if they exist, but it would change very little from how targets are used, and thus we would not disrupt any existing workflows.

Just a thought.

**Comment by dansondergaard** [comment]
I never wanted to make targets really immutable (impossible in Python anyway), I just want to treat them mostly as immutable :)

I agree that it'd be inconsistent if the target wasn't returned from `from_template`, so I left it there.

Also, I can't come up with a case where the one would use a template *and* provide inputs/outputs to the target directly. The whole purpose of using a template is to let the template function deal with setting things like this from the arguments. At least, I've never overridden inputs/outputs from a template.

Abusing the inputs and outputs properties is definitely possible. I just don't like the idea of "validating" a target. However, if that's what it takes I won't complain. However, I think it's just another (much less ugly) hack that doesn't really reduce the complexity.

**Comment by mailund** [comment]
I'm not entirely against the idea of having a `from_template()` method -- I would probably prefer `target_from_template()` as the method name, but that is a minor issue. It is definitely better than the solution I hacked up yesterday.

It will break a few workflows, but people *have* been warned that this could happen until we release v1.0. I would just prefer that we didn't break them if we can avoid it.

We don't validate all of a target at construction time, though, so it is not as if having a `validate()` method entirely changes the design. The backends also validate targets to make sure that the options they get are known to them -- at least, if you provide an option to the slurm backend that it doesn't understand it will throw an exception.

By validating targets after they are constructed, though, we do add some flexibility to them. It will be possible to extend gwf in various ways through functions that manipulate targets. I could see how that could be useful from playing around with language extensions earlier. You can still do this even if we make them "mostly immutable" but it would be much easier to screw it up.

I'm not entirely sure what I prefer. In some ways, I really like the idea that once a target is added to the workflow it is immutable, but on the other hand, I also like to think of targets as objects that can be manipulated.

Of course, one way to achieve both would be to separate creating targets from adding them to the workflow. Right now, we add them to the workflow and then modify them. We could separate the two operations so we would write

```python
foobar = Target(inputs = [], outputs = ["foobar"], cores = 8) << """
some script
"""
gwf.add_taget("foobar", foobar)
```

and then validate the target in `add_target()`.

Splitting the target creation from adding it to the workflow of course opens a new source of confusion for users, so we probably want to keep the `gwf.target()` method as well, so we could have

```python
gwf.target("foobar", inputs = ..., outputs ...) << "foobar"
```

as we have now,

```python
gwf.from_template("foobar", foobartemplate())
```

as you suggest, and then 

```python
gwf.add_target("foobar", foobar)
```

(perhaps without the name specified there but in the Target creation).

Of course, now we have *three* ways of adding targets, and that causes confusion as well, so I don't know...

**Comment by mailund** [comment]
With the three adding methods, I would only let `gwf.target()` return the target, though. Just to make it clear that two of them adds targets that should not be modified later.

**Comment by dansondergaard** [comment]
> I'm not entirely against the idea of having a from_template() method -- I would probably prefer target_from_template() as the method name, but that is a minor issue. It is definitely better than the solution I hacked up yesterday.

Both names are fine by me!

> The backends also validate targets to make sure that the options they get are known to them -- at least, if you provide an option to the slurm backend that it doesn't understand it will throw an exception.

Soon, backends won't throw an exception, they will just warn the user that the backend doesn't support the given option. I don't really see it as validation of the target, it's just the backend figuring out which options it can use - but I get your point :)

> Of course, one way to achieve both would be to separate creating targets from adding them to the workflow.

That's inconvenient (to put it mildly).

I think we should keep it simple and either go with your (better) hack or the `target_from_template()` method... or completely different solution. What does @birc-aeh think?

**Comment by birc-aeh** [comment]
Delay the check as necessary. Just like we check for cycles and so on later in the process it would be easy to report that `target 'MyTarget' has no inputs` after the full graph has been built.

I think the real usability issues are probably more towards the UI side of things - especially when there are thousands of jobs.

**Comment by dansondergaard** [comment]
@birc-aeh As always, clear and precise!

Ok, I'll go with delayed validation of the target in `Graph`, especially @mailund can make it prettier using the properties :)

Also, I agree that status on large workflows is a huge challenge. If @mailund takes care of this I'll look into querying workflows and print useful info. I have a few ideas for this.

**Comment by mailund** [comment]
I'll have a look over the weekend.

**Comment by dansondergaard** [comment]
@mailund Any news on this? :)

**Comment by micknudsen** [comment]
Any updates? I thought that a temporary fix had been implemented, but it appears to be broken in the current master.

**Comment by dansondergaard** [comment]
A temporary fix (which I'm considering to also make the final fix) is to just set ``inputs=[], outputs=[]`` in the target. They will be overriden when assigning the template.

**Comment by micknudsen** [comment]
Hopefully not something the user will have to do explicitly. You will set that behind the scenes (in the code), right?

**Comment by dansondergaard** [comment]
The fix that Mailund wrote (and the one he was planning) would do this implicitly, yes.

However, there's a reason that we decided that `inputs` and `outputs` should be set explicitly when creating targets: the user may unintentionally forget to set inputs/outputs which can cause various confusing problems. When the user is forced to set inputs/outputs to the empty list, the intention is very explicit.

So setting them when using a template seems like a fair requirement too, even though it is annoying for the user.

I still kinda prefer this solution:

```python
gwf.target('MyTarget', my_template(foo='hello', bar='world'))
```

which would raise an error if the template didn't declare `inputs` and `outputs`. Similarly, creating a target the usual way, but not specifying `inputs` and `outputs` would raise an error.

IMHO, `from_template` is even more elegant and explicit. It clearly states the intention of the user and separates two concerns: adding a normal target to the workflow and adding a target made from a template to the workflow.

**Comment by micknudsen** [comment]
Just tried this test workflow (on Mac laptop using local backend with 8 workers):

```
from gwf import Workflow

gwf = Workflow()

def my_template(foo):

    options = { 'inputs': [], 'outputs': ['bar.txt']}

    spec = f'''
    echo {foo} > bar.txt
    '''

    return options, spec

gwf.target('test', inputs=[], outputs=[]) << my_template(foo='baz')
```

Why this warning?

```
(test_gwf) [micknudsen@work:~/Temp]$ gwf run
Scheduling target test.
Submitting target test.
/Users/micknudsen/Temp/gwf/gwf/backends/base.py:55: UserWarning: Backend does not support option inputs used in test. Option will be ignored.
  option_name, target.name
/Users/micknudsen/Temp/gwf/gwf/backends/base.py:55: UserWarning: Backend does not support option outputs used in test. Option will be ignored.
  option_name, target.name
```

**Comment by mailund** [comment]
This is the reason it was difficult to implement a validation function. If we keep `inputs` and `outputs` in the `options` variable, we run into problems with the backends that validate their options. I think it might be worse than that if you don't use my fix. I am not sure the `inputs` and `outputs` of the target will get set in your example. I think `gwf` will just use the arguments to the `target` since these will always overrule any template arguments -- which is the desired behaviour.

I am beginning to like @dansondergaard's suggest more and more. We really *do* want people to explicitly specify `inputs` and `outputs` to avoid forgetting about them. Having them be empty by default is a real source of error when combined with taking generic keyword arguments.

**Comment by dansondergaard** [comment]
@micknudsen Yeah, that's a bug.

Everything would be simpler and make more sense if (1) we had something like `from_template` and (2) templates were of the form `(inputs, outputs, options, spec)`. Inputs and outputs are not options, they required for defining a target and thus they should not be specified in the options dictionary.

What do you think, @mailund, should we go with this approach?

**Comment by mailund** [comment]
I would be happy with this solution, yes.

**Comment by dansondergaard** [comment]
Do you want to code it up, @mailund? Would be cool to have it done pretty soon so we can get a beta 4 out. I'll try to write some docs during the weekend.

**Comment by mailund** [comment]
I can do it tomorrow.

**Comment by dansondergaard** [comment]
Just to coordinate:

* the method will be called `target_from_template`,
* the method will accept three arguments: name of the target, a template of the form `(inputs, outputs, options, spec)`, and an optional argument with a dict that overrides options set by the template (I suggest that we name this argument `override`.

Example:

```python
gwf.target_from_template('Foo', my_template(x=1, y=2), override={'walltime': '16:00:00'})
```

Do you agree? :)

**Comment by mailund** [comment]
I *almost* agree. The one thing I disagree with is the `override` option. For the `gwf.target` method, we use keyword arguments to provide options. For consistency, I think we should use the same here. Alternatively, we need to explicitly provide a `dict` with the `target` method, but I think I like the keyword arguments more.

There isn't really any difference between the two solutions except for syntax. None is safer than the other and both are just standard Python--and once inside the method, we need to treat both kinds of providing options through a `dict`--so it is purely an aesthetics choice, but I prefer the keyword version as it saves a few keystrokes.

I will implement the `target_from_template` that way now and make a pull request; then we can discuss.

I won't update the template creation function--if we even have that one still--until we've discussed it. We might want to get rid of it completely now. It was a hack anyway that the new function interface is much superior to.

I *will* get rid of the option of providing a template to the shift operator for targets, though.

Do all agree with this?

**Comment by dansondergaard** [comment]
I agree. The reason that I suggested the `override` argument is that it make it explicit that those arguments will override template-provided options (again, explicit over implicit). However, I'm happy with both solutions.

**Comment by mailund** [comment]
Resolved with PR #157
